crates/node/src/metadata.rs

```diff
@@ -0,0 +1,409 @@
+#[derive(Clone)]
+pub struct MetadataWriter {
+    sender: CommandSender,
+}
+
+impl MetadataWriter {
+    fn new(sender: CommandSender) -> Self {
+        Self { sender }
+    }
+
+    // Returns when the nodes configuration update is performed.
+    pub async fn update(&self, value: impl Into<MetadataContainer>) -> Result<(), ShutdownError> {
+        let (callback, recv) = oneshot::channel();
+        let o = self
+            .sender
+            .send(Command::UpdateMetadata(value.into(), Some(callback)));
+        if o.is_ok() {
+            let _ = recv.await;
+            Ok(())
+        } else {
+            Err(ShutdownError)
+        }
+    }
+
+    // Fire and forget update
+    pub fn submit(&self, value: impl Into<MetadataContainer>) {
+        // Ignore the error, task-center takes care of safely shutting down the
+        // system if metadata manager failed
+        let _ = self
+            .sender
+            .send(Command::UpdateMetadata(value.into(), None));
+    }
+}
+
+/// What is metadata manager?
+///
+/// MetadataManager is a long-running task that monitors shared metadata needed by
+/// services running on this node. It acts as the authority for updating the cached
+/// metadata. It can also perform other tasks by running sub tasks as needed.
+///
+/// Those include but not limited to:
+/// - Syncing schema metadata, logs, nodes configuration with admin servers.
+/// - Accepts adhoc requests from system components that might have observed higher
+/// metadata version through other means. Metadata manager takes note and schedules a
+/// sync so that we don't end up with thundering herd by direct metadata update
+/// requests from components
+///
+/// Metadata to be managed by MetadataManager:
+/// - Bifrost's log metadata
+/// - Schema metadata
+/// - NodesConfiguration
+/// - Partition table
+pub struct MetadataManager {
+    self_sender: CommandSender,
+    inner: Arc<MetadataInner>,
+    inbound: CommandReceiver,
+}
+
+impl MetadataManager {
+    pub fn build() -> Self {
+        let (self_sender, inbound) = tokio::sync::mpsc::unbounded_channel();
+
+        Self {
+            inner: Arc::new(MetadataInner::default()),
+            inbound,
+            self_sender,
+        }
+    }
+
+    pub fn metadata(&self) -> Metadata {
+        Metadata::new(self.inner.clone(), self.self_sender.clone())
+    }
+
+    pub fn writer(&self) -> MetadataWriter {
+        MetadataWriter::new(self.self_sender.clone())
+    }
+
+    /// Start and wait for shutdown signal.
+    pub async fn run(mut self /*, network_sender: NetworkSender*/) -> anyhow::Result<()> {
+        info!("Metadata manager started");
+
+        loop {
+            tokio::select! {
+            biased;
+            _ = cancellation_watcher() => {
+                info!("Metadata manager stopped");
+                break;
+            }
+            Some(cmd) = self.inbound.recv() => {
+                self.handle_command(cmd)
+            }
```

Comment:
Indentation looks a bit off. Might be GitHub.

Reply:
Macros and rustfmt don't like each other :)

Will indent.